conn_test.go

```diff
@@ -2064,6 +2064,131 @@ func TestServerNameIndicationExtension(t *testing.T) {
 	}
 }
 
+func TestALPNExtension(t *testing.T) {
+	// Limit runtime in case of deadlocks
+	lim := test.TimeOut(time.Second * 20)
+	defer lim.Stop()
+
+	// Check for leaking routines
+	report := test.CheckRoutines(t)
+	defer report()
+
+	for _, test := range []struct {
+		Name                   string
+		ClientProtocolNameList []string
+		ServerProtocolNameList []string
+		Expected               string
+	}{
+		{
+			Name:                   "Negotiate a protocol",
+			ClientProtocolNameList: []string{"http/1.1", "spd/1"},
+			ServerProtocolNameList: []string{"spd/1"},
+			Expected:               "spd/1",
+		},
+		{
+			Name:                   "Server doesn't support",
+			ClientProtocolNameList: []string{"http/1.1", "spd/1"},
+			ServerProtocolNameList: []string{},
+			Expected:               "",
+		},
```

Comment:
I think we should add a third test here. Multiple client and server protocols, where their order doesn't match and not all protocols overlap.

Something like:

```
ClientProtocolNameList: []string{"http/1.1", "spd/1", "http/3"},
ServerProtocolNameList: []string{"ssh/2", "http/3", "spd/1"},
Expected: "http/3"
```

Reply:
Done